pesudocode/voxel_grid_structure_2.hpp

```diff
@@ -0,0 +1,96 @@
+#include <iostream>
+#include <vector>
+#include <unordered_map>
+
+//  base class inherits class specialized
+
+class PointXYZ{
+    public:
+    PointXYZ(){
+    }
+};
+
+template <typename PointT>
+class Filter{
+    public:
+    Filter(){}
+};
+
+// new base class
+
+// in order to initialize like this:
+// GridFilter<VoxelGridT<PointT>>
+template <typename PointT>
+class GridFilter;
+
+template <template <typename> class Grid, typename PointT>
+class GridFilter<Grid<PointT>>: public Filter<PointT>, protected Grid<PointT>{
```

Comment:
Assumption of a templated class is not good. How about:
```cpp
template <class Grid, typename PointT=get_point_t<Grid>>
class GridFilter<Grid>: public Filter<PointT>, protected Grid
```

get_point could do a number of possible strategies:
* check for a typedef inside the class
* check the 1st templated parameter
* allow user to override the struct `get_point<Grid>` (since get_point_t<Grid> = get_point<Grid>::type)

Reply:
Looks like checking a typedef/using would be a better choice. I don't know a straightforward way to also check the mandatory functions/methods before returning the point type from checking the template parameter

I think we have to check the mandatory stuff beforehand, otherwise the compiler will complain a lot of unintuitive errors if the provided methods is incompatible

Reply:
Resolving this since the newer draft of the API is nicer